### ISSUE TITLE
[User] PrefetchVirtualMemory not found in KERNEL32.dll

### ISSUE THREAD

**Issue by NBcobo** [LostRuins/koboldcpp#68]
Hi, I'm running Windows 7 and got the above error message. Similar issue can be found here: https://github.com/ggerganov/llama.cpp/issues/894
And one possible solution might be in: 
https://github.com/ggerganov/llama.cpp/pull/890
Is it possible to modify your program to run on Windows 7.
Thank you.

**Comment by LostRuins** [comment]
Hm if it's merged already 4 days ago it should be in the latest version. I will have to check

**Comment by anzz1** [comment]
The PR https://github.com/ggerganov/llama.cpp/pull/890 brought the https://github.com/ggerganov/llama.cpp/issues/894 issue, and the solution is in the comment https://github.com/ggerganov/llama.cpp/pull/890#issuecomment-1503354856 , however the fix was posted *after* the PR was merged and thus not included in the PR.

**Comment by LostRuins** [comment]
Is this issue resolved in the latest release? I have already merged the upstream changes in v1.10

**Comment by anzz1** [comment]
> Is this issue resolved in the latest release? I have already merged the upstream changes in v1.10

Of llama.cpp? Nop.

**Comment by LostRuins** [comment]
Ah okay. Is there an existing PR or issue for it in the upstream llama.cpp repo? I'd rather merge changes from there than modify it on my own copy

**Comment by anzz1** [comment]
Yep [#894](https://github.com/ggerganov/llama.cpp/issues/894) is open

**Comment by MasterDimmy** [comment]
same issue
win7 x64 not worked

**Comment by NBcobo** [comment]
This only affects Windows 7, which they may feel not have to support.

**Comment by LostRuins** [comment]
@NBcobo I did a quick test and that code didn't compile. However, I did notice that PrefetchVirtualMemory seems to only be triggered when running with mmap. Therefore, on Windows 7 can you please try the latest version and run it with the flag `--nommap` and I think it should work.

**Comment by NBcobo** [comment]
Just tried to run command line "koboldcpp.exe --nommap" and selected one ggml .bin file manually. Here is what I got:
![image](https://user-images.githubusercontent.com/130814579/236269652-b9c53424-3a97-4ff8-8481-09c5a2f13c16.png)

**Comment by anzz1** [comment]
```sh
cd koboldcpp
wget https://github.com/LostRuins/koboldcpp/files/11405936/llama-util.patch
git apply -p1 < ./llama-util.patch
```
[llama-util.patch](https://github.com/LostRuins/koboldcpp/files/11405936/llama-util.patch)

**Comment by anzz1** [comment]
You can use mmap on Win7 just fine, you just can't prefetch (well, you can, just not with that winapi function). Using mmap with prefetch pretty much defeats the point of using mmap anyway, unless you're using multiple llama processes concurrently.

With mmap, `--mlock`, without prefetch: Loads a "view" of the file (model) to memory, contents stay on disk until accessed. Starts with low memory usage, which gradually grows as more parts of the model are accessed. Instant startup with the tradeoff of reduced generation speed, as the model is loaded on the go rather than preloaded fully at the beginning.

Without mmap, `--no-mmap`, regular fopen/fread: Loads the whole model to memory, disk is not accessed after startup. As the whole model is loaded from disk to memory before anything is done, starts at maximum memory usage which doesn't grow any more. Slow startup, but fast generation speed.

With mmap, `--mlock`, with prefetch: Loads a "view" of the model, but also preloads the contents. For all intents and purposes the same as fopen/fread, with the difference that multiple llama processes can share the same shared memory object.

With mmap, no lock, prefetch/no prefect: Otherwise the prefetch/no prefetch rules apply, but without locking the OS will release parts of the memory after use to be loaded from disk once again later. Greatly reduced memory use, with greatly reduced generation speed. When performance is of concern, like the case is with llama, you'd never want this.

If you have enough memory and aren't straddling near capacity, and using a single llama process, mmap-with-prefetch is pretty much the same as --no-mmap. However, if you are near your memory limit and your OS needs to start paging stuff out of memory to not crash, no-mmap would need to write the memory contents to the system pagefile and later read from there, which is a bad thing. In this case mmap would be better, as it can only drop the memory and read from the model file, no need to write to pagefile. However, the problem is that your OS also knows this, so mmap'ed memory takes lower precedence than "regular" one. So the OS is more likely to drop the memory contents from a process using mmap than something else which is not. So llama with mmap would take a L instead of some other process running on your PC in this case, while llama with no-mmap would have a higher priority and thus more likely to keep it's memory contents available. In any case your generation speed will tank if you're near memory capacity, and you don't want that.

**TL;DR;**
- Don't be near your memory capacity while running the models, choose a model which fits nicely in your memory with some space to spare.
- Win7, use `--no-mmap` if you want preloaded model, slow startup but fast generation. Don't use if you want fast startup but slower generation.
- Win8+ doesn't really matter as prefetch is used, mmap with mlock is basically same as no-mmap.
- Both: Always use either `--no-mmap` or `--mlock` , don't use mmap without `--mlock`.

**Comment by LostRuins** [comment]
@anzz1 how about submitting that change to the upstream repo so win7 users don't have to manually patch it in?

**Comment by anzz1** [comment]
I am no longer a collaborator on the upstream project, but anyone can feel free to submit it if they want, no credit necessary.

**Comment by NBcobo** [comment]
Here is what I got
![image](https://user-images.githubusercontent.com/130814579/236590485-bd88898a-8263-41f7-ae69-1e3ab4e10217.png)
Thank you  so much for the explanation and code. I may just upgrade my OS to Windows 10.

**Comment by LostRuins** [comment]
Hey @NBcobo I've gone ahead and built a special windows 7 compatible exe without prefetch so you can try. 

Please let me know if this works for you: https://github.com/LostRuins/koboldcpp/releases/download/v1.19/koboldcpp_win7.exe

**Comment by NBcobo** [comment]
Here is what I got. I think it's the similar issue I encountered with PyInstaller. It's time for me to move on beyond Windows 7. Thank you so much for your time and effort!
![image](https://user-images.githubusercontent.com/130814579/236630857-4cb5b03d-3a2b-4039-a34c-c99251da0517.png)

**Comment by NBcobo** [comment]
After I moved to Windows 10, I found out that my CPU only support AVX, not AVX2. So, with the option --noavx2, the codes run fine. For those who are still with Windows 7, you may want to check your CPU first. For the Python 3.9 and above supporting issue, you may want to check this comment https://github.com/LostRuins/koboldcpp/issues/147#issuecomment-1537159709

**Comment by LostRuins** [comment]
For anyone still trying to get it working on windows 7, there is now a special test build that has been found to work on windows 7. You can try it here: https://github.com/LostRuins/koboldcpp/releases/download/v1.25/koboldcpp_win7_test.exe

**Comment by feakk** [comment]
@LostRuins I didn't see this mentioned anywhere, so confirming that [koboldcpp_win7_test.exe](https://github.com/LostRuins/koboldcpp/releases/download/v1.25.1/koboldcpp_win7_test.exe) works on Windows 7 (whereas v1.19/koboldcpp_win7.exe, and other version of llama and koboldcpp don't). Thanks for the extra support, as it looks like [#894](https://github.com/ggerganov/llama.cpp/issues/894) needs a gentle push for traction support.
